subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/PathUtil.java

```diff
@@ -41,6 +41,11 @@
  * - Finally, if both segments are the same ignoring case and have the same length, the case-sensitive comparison is used.
  *
  * For all methods operating on a list of paths, the paths must not have a common, non-trivial prefix.
+ *
+ * Most of the methods take an absolutePath and an offset within the path.
+ * This is a performance optimization so that we don't need to call {@link String#substring(int)} to obtain a relative path from an absolute path.
+ * The absolutePath argument of the method is supposed to be an absolute path and the offset determines the relative path to the examined location.
+ * For example, filePath = /some/foo/bar with offset 6 determines the relative path foo/bar from the location /some.
```

Comment:
It would be good to state that `offset` should always point after the separator, not to the separator.

Reply:
Maybe add a `// TODO` about introducing a separate type to capture all this?

Reply:
Would be good to define "segment" and "prefix" here, too.

Reply:
Done.